### PR TITLE
[wip/rfc] Guard filter hook

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134311
* #134354
* #134272

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Review by youkaichao — COMMENTED** [review]
this is a great effort!

File: torch/_dynamo/hooks.py [hunk 1]
```diff
@@ -10,3 +11,60 @@
...
+
+
+def skip_guards_on_nn_module_attributes(guards: GuardsSet) -> GuardsSet:
+    # Skip guards on nn module attributes. For deep models, a lot of time can be spent on nn module attribute checks
+    from torch._dynamo.guards import guards_log
+
+    new_guards = GuardsSet()
+
+    for g in guards:
+        if g.source.is_unspecialized_nn_module() or g.source.is_specialized_nn_module():
```
Comment (youkaichao) [inline torch/_dynamo/hooks.py:58]: I'm interested in what functionality we have here, i.e. what is the structure and information I can get from the guard, that we can use to filter.

Comment (anijain2305) [reply torch/_dynamo/hooks.py:58]: Yes, so a lot of the constructs here are Dynamo internals and there is no other easy way than to understand these if you want more control.

The Guard data structure is here - https://github.com/pytorch/pytorch/blob/main/torch/_guards.py#L160

So it has everything needed - originating_source (where is this guard coming from), guard_source (is this coming from nn module - specialized vs unspeicalised), name (how to access the object from f_locals) etc.

File: torch/_dynamo/guards.py [hunk 2]
```diff
@@ -2203,6 +2212,9 @@ def cleanup_builder(weak_b):
         # Break retain cycle. See test_release_input_memory
         w_builder = weakref.ref(builder, cleanup_builder)
 
+        if guard_filter_fn:
+            guards = guard_filter_fn(guards)
```
Comment (youkaichao) [inline torch/_dynamo/guards.py:2216]: I understand `guard_filter_fn` is used here. But how about `symbolic_guard_filter_fn`? where is it used?

And, if I use `guard_filter_fn`, will it be applied to symbolic guards as well?

Comment (anijain2305) [reply torch/_dynamo/guards.py:2216]: No, that's the clunky part. The symbolic guards come from a different place. And therefore we need 2 different hooks.

**Pull request closed without merging**